### PR TITLE
feat: bound partition key size

### LINKED ISSUE #7900: Length limit on partition keys
From https://github.com/influxdata/influxdb_iox/pull/7893#discussion_r1211867171:

* Truncate each partition part to at most 200 bytes (NOT chars!)
* I will include an explicit truncation marker (#) as the last character of the truncated string (which might be <200 chars, because unicode - él es molesto)
* I will "reserve" / encode any other occurrence of the # character like the other special chars
* Limit the number of template parts to 8

Therefore the max partition key length will be 200*8 = 1,600 bytes. I will return an error for the write if it exceeds any length limit. I will also return an error when creating a custom template if it contains > 8 parts, so the user has a better time if they exceed it.

### PR DESCRIPTION
**Pull request by domodwyer**
This PR bounds both the individual parts within a partition key, and the number of parts allowed in a key - this effectively bounds the overall size of the key to 1,607 bytes (1.57 KiB).

The reversal function that reconstructs the tag column values from the partition key now emits a tuple of `(column_name, column_value_enum)` which encodes the valid operations (equality/prefix-match) into the type system.

Closes #7900.

---

* feat: truncate partition key parts at 200 bytes (050093df1)
      
      This commit ensures all partition key parts are less than or equal to
      200 bytes long.
      
      If a string exceeds the 200 byte limit, it is truncated (avoiding
      splitting unicode code-points or graphemes) and then a single "#"
      sentinel value is appended. When reversed from the string, these column
      values are indicated to be suitable for prefix-matching only - a
      property that is encoded into the type system.
      
      This commit takes a conservative approach of not splitting graphemes as
      outlined in the module documentation, but this could be relaxed in the
      future if needed.

* refactor: expose partition part count (39c22a2c2)
      
      Allow a table template override to report the number of template parts
      within it.
      
      This ignores the lint wanting an "is_empty()" method too, because it's
      misleading and redundant - a template MUST never be empty.

* feat: restrict partition template length (fc49b3ec1)
      
      Partition templates should not contain more than 8 parts, which when
      combined with a per-part byte limit, bounds the maximum size of a
      partition key.
      
      This commit causes the router to refuse to service a write request that
      contains > 8 parts in the template - this causes a panic, as it's a
      broken system invariant and should be an unreachable state. Templates
      are pre-validated at creation time to contain no more than 8 parts, and
      are immutable:
      
          https://github.com/influxdata/influxdb_iox/pull/7930

### CODE REVIEW

File: data_types/src/partition_template.rs [hunk 1]
```diff
@@ -47,14 +47,68 @@
...
+//! option - we can't easily do the reverse!
+//!
+//! ## Part Limit & Maximum Key Size
+//!
+//! The number of parts in a partition template is limited to 8
+//! ([`MAXIMUM_NUMBER_OF_TEMPLATE_PARTS`]), validated at creation time.
+//!
+//! Together with the above value truncation, this bounds the maximum length of
+//! a partition key to 1,607 bytes (1.57 KiB).
+//!
```
Comment (domodwyer) [inline data_types/src/partition_template.rs:102]: This block describes how partition key part truncation is applied, particularly the considerations made for unicode strings.

TL;DR: a truncated value is at most 200 bytes long, but has a variable length below that - it's possible to see a truncated value that is 195 bytes long. This is an implementation detail and encoded into the type system; the user of these types is not expected to make any special effort with these truncated strings.

File: data_types/src/partition_template.rs [hunk 2]
```diff
@@ -366,6 +442,54 @@ mod serialization {
...
+
+    /// The inner value is a variable length prefix of the input column value.
+    ///
+    /// The string value is always guaranteed to be valid UTF-8.
+    ///
+    /// Attempting to equality match this variant against a string will always
+    /// be false - use [`ColumnValue::is_prefix_match_of()`] to prefix match
+    /// instead.
+    Prefix(Cow<'a, str>),
+}
```
Comment (domodwyer) [inline data_types/src/partition_template.rs:461]: A rebuilt column value is now emitted wrapped in this enum.

File: data_types/src/partition_template.rs [hunk 3]
```diff
@@ -366,6 +442,54 @@ mod serialization {
...
+where
+    T: AsRef<str>,
+{
+    fn eq(&self, other: &T) -> bool {
+        match self {
+            ColumnValue::Identity(v) => other.as_ref().eq(v.as_ref()),
+            ColumnValue::Prefix(_) => false,
+        }
+    }
+}
```
Comment (domodwyer) [inline data_types/src/partition_template.rs:491]: This type can only be used for equality matching of strings when it is the "identity" variant...

File: data_types/src/partition_template.rs [hunk 4]
```diff
@@ -366,6 +442,54 @@ mod serialization {
...
+    where
+        T: AsRef<[u8]>,
+    {
+        let this = match self {
+            ColumnValue::Identity(v) => v.as_bytes(),
+            ColumnValue::Prefix(v) => v.as_bytes(),
+        };
+
+        other.as_ref().starts_with(this)
+    }
```
Comment (domodwyer) [inline data_types/src/partition_template.rs:478]: ...however both variants can be used to prefix-match against a string.

File: mutable_batch/src/payload/partition.rs [hunk 5]
```diff
@@ -147,13 +153,49 @@ impl<'a> Template<'a> {
...
+                })
+                .for_each(|v| buf.push_str(v.as_ref()));
+
+            // Append the truncation marker.
+            buf.push(PARTITION_KEY_PART_TRUNCATED);
+
+            assert!(buf.len() <= PARTITION_KEY_MAX_PART_LEN);
+
+            Cow::Owned(buf)
+        }
```
Comment (domodwyer) [inline mutable_batch/src/payload/partition.rs:197]: This is the truncation logic, which accounts for both unicode code-points and graphemes when splitting.

This encoding is optimised for <200 character values, expecting >200 character values to be the unusual case.

**Review by savage-engineer — APPROVED** [review]
🎩

File: data_types/src/partition_template.rs [hunk 6]
```diff
@@ -641,7 +770,40 @@ mod tests {
...
+
+    test_build_column_values!(
+        unicode_code_point_prefix,
+        template = [
+            TemplatePart::TimeFormat("%Y"),
+            TemplatePart::TagValue("a"),
+            TemplatePart::TagValue("b"),
+        ],
+        partition_key = "2023|%28%E3%83%8E%E0%B2%A0%E7%9B%8A%E0%B2%A0%29%E3%83%8E%E5%BD%A1%E2%94%BB%E2%94%81%E2%94%BB#|!",
+        want = [("a", prefix("(ノಠ益ಠ)ノ彡┻━┻"))]
```
Comment (carols10cents) [inline data_types/src/partition_template.rs:827]: 😹

File: mutable_batch/src/payload/partition.rs [hunk 7]
```diff
@@ -762,20 +1077,34 @@ mod tests {
...
+                assert_eq!(got.0, want.0, "column names differ");
+
+                match got.1 {
+                    ColumnValue::Identity(_) => {
+                        // An identity is both equal to, and a prefix of, the
+                        // original value.
+                        assert_eq!(got.1, want.1, "identity values differ");
+                        assert!(got.1.is_prefix_match_of(&want.1), "prefix mismatch");
+                    },
+                    ColumnValue::Prefix(_) => assert!(got.1.is_prefix_match_of(&want.1), "prefix mismatch"),
```
Comment (carols10cents) [inline mutable_batch/src/payload/partition.rs:1105]: Thinking about debugging these tests if they fail, for the two `assert!`s here it'd be nice if the args to `is_prefix_match_of` were included in the failure message, something like:

```suggestion
                        assert!(got.1.is_prefix_match_of(&want.1), "prefix mismatch; {} is not a prefix of {}", got.1, want.1);
                    },
                    ColumnValue::Prefix(_) => assert!(got.1.is_prefix_match_of(&want.1), "prefix mismatch; {} is not a prefix of {}", got.1, want.1),
```

File: data_types/src/partition_template.rs [hunk 8]
```diff
@@ -218,6 +285,15 @@ impl TablePartitionTemplateOverride {
...
 
+    /// Returns the number of parts in this template.
+    #[allow(clippy::len_without_is_empty)] // Senseless - there must always be >0 parts.
+    pub fn len(&self) -> usize {
+        self.0
+            .as_ref()
+            .map(|v| v.inner().parts.len())
+            .unwrap_or_default()
+    }
+
```
Comment (carols10cents) [inline data_types/src/partition_template.rs:296]: As written, this function *will* currently return 0 in this case (this test currently passes):

```
    #[test]
    fn table_partition_template_none() {
        let ns = NamespacePartitionTemplateOverride::default();
        let t = TablePartitionTemplateOverride::try_new(None, &ns).unwrap();

        assert_eq!(t.len(), 0);
    }
```

Which seems to contradict the comment "there must always be >0 parts"?

If `len` instead uses `parts`, then it will get the default partition template and indeed never return 0 (at the cost of iterating through all the parts):

```suggestion
    /// Returns the number of parts in this template.
    #[allow(clippy::len_without_is_empty)] // Senseless - there must always be >0 parts.
    pub fn len(&self) -> usize {
        self.parts().count()
    }
```

That is, the test that will pass now is:

```
    #[test]
    fn table_partition_template_none() {
        let ns = NamespacePartitionTemplateOverride::default();
        let t = TablePartitionTemplateOverride::try_new(None, &ns).unwrap();

        assert_eq!(t.len(), 1);
    }
```

because the hardcoded default template has 1 part.

I see the `len` method is only used to check the invariant of the upper bound of number of parts, so I'm not sure how important this is. If you want to keep the code the way it is, then I think this case/assumption should be documented at least?

File: mutable_batch/src/payload/partition.rs [hunk 9]
```diff
@@ -49,6 +51,14 @@ pub fn partition_batch<'a>(
     batch: &'a MutableBatch,
     template: &'a TablePartitionTemplateOverride,
 ) -> impl Iterator<Item = (Result<String, PartitionKeyError>, Range<usize>)> + 'a {
+    let parts = template.len();
+    if parts > MAXIMUM_NUMBER_OF_TEMPLATE_PARTS {
+        panic!(
+            "partition template contains {}, which exceeds the maximum of {}",
```
Comment (carols10cents) [inline mutable_batch/src/payload/partition.rs:57]: For clarity for the person theoretically getting paged over this, I'd change the message to include the word "parts" so that someone doesn't start wondering why a literal 9 isn't allowed in a partition template:

```suggestion
            "partition template contains {} parts, which exceeds the maximum of {} parts",
```

**Review by carols10cents — CHANGES_REQUESTED** [review]
A bunch of nits; I'd definitely like to see the `len` method changed or its documentation changed before merging.

File: data_types/src/partition_template.rs [hunk 10]
```diff
@@ -424,12 +548,27 @@ pub fn build_column_values<'a>(
...
+                .last()
+                .map(|v| *v == PARTITION_KEY_PART_TRUNCATED as u8)
+                .unwrap_or_default()
+            {
+                // Remove the truncation marker.
+                let len = decoded.len() - 1;
+                return (
+                    name,
+                    ColumnValue::Prefix(Cow::Owned(decoded[..len].to_string())),
+                );
```
[the remark this answers is not in the document]
Comment (domodwyer) [reply data_types/src/partition_template.rs:568]: I like it!

File: mutable_batch/src/payload/partition.rs [hunk 11]
```diff
@@ -762,20 +1077,34 @@ mod tests {
...
                     Some((col_name, tag_values.get(col_name).unwrap().to_string()))
                 }
                 _ => None,
-            }).collect::<Vec<_>>();
-
-            assert_eq!(reversed, want_reversed);
+            }).collect();
+
+            for (idx, want) in want_reversed.iter().enumerate() {
+                let got = &reversed[idx];
```
[the remark this answers is not in the document]
Comment (domodwyer) [reply mutable_batch/src/payload/partition.rs:1094]: I like it too! Indexing I think will always add an implicit range check too (unless it can be proven to exist some other way), so this will probably be faster too!

**Review by carols10cents — APPROVED** [review]

**Comment by carols10cents** [comment]
I fixed all my issues with this PR, but @domodwyer should review the commits I added before merging :)

**Comment by domodwyer** [comment]
Thank you @carols10cents! Super helpful 🙏
